engine/runtime/src/main/java/org/enso/interpreter/runtime/SerializationPool.java

```diff
@@ -0,0 +1,187 @@
+package org.enso.interpreter.runtime;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import org.enso.pkg.QualifiedName;
+
+final class SerializationPool {
+  private final TruffleCompilerContext context;
+
+  /**
+   * A set of the modules that are currently being serialized.
+   *
+   * <p>This set is accessed concurrently. This is safe as it is backed by a [[ConcurrentHashMap]]
+   * and is wrapped with the scala [[mutable.Set]] interface.
+   */
+  private final Map<QualifiedName, Boolean> isSerializing = new ConcurrentHashMap<>();
+
+  /**
+   * A map of the modules awaiting serialization to their associated tasks
+   *
+   * <p>This map is accessed concurrently.
+   */
+  private final Map<QualifiedName, Future<?>> isWaitingForSerialization = new ConcurrentHashMap<>();
+
+  /** The thread pool that handles serialization. */
+  private final ExecutorService pool;
+
+  /** all associated threads */
+  private final Set<Thread> threads = Collections.synchronizedSet(new HashSet<>());
+
+  SerializationPool(TruffleCompilerContext context) {
+    this.context = context;
+    this.pool =
+        Executors.newSingleThreadExecutor(
+            (r) -> {
+              var t = context.createSystemThread(r);
+              t.setName("SerializationPool background thread");
+              threads.add(t);
+              return t;
+            });
+  }
+
+  void prestartAllCoreThreads() {}
+
+  /**
+   * @return `true` if there are remaining serialization jobs, `false` otherwise
+   */
+  private boolean hasJobsRemaining() {
+    synchronized (isWaitingForSerialization) {
+      return !isWaitingForSerialization.isEmpty() || !isSerializing.isEmpty();
+    }
+  }
+
+  /**
+   * Performs shutdown actions for the serialization manager.
+   *
+   * @param waitForPendingJobCompletion whether or not shutdown should wait for pending
+   *     serialization jobs
+   */
+  void shutdown(boolean waitForPendingJobCompletion) throws InterruptedException {
+    if (!pool.isShutdown()) {
+      if (waitForPendingJobCompletion && this.hasJobsRemaining()) {
+        int waitingCount;
+        int jobCount;
+        synchronized (isWaitingForSerialization) {
+          waitingCount = isWaitingForSerialization.size();
+          jobCount = waitingCount + isSerializing.size();
+        }
+        context.logSerializationManager(
+            Level.FINE, "Waiting for #{0} serialization jobs to complete.", jobCount);
+
+        // Bound the waiting loop
+        int maxCount = 60;
+        int counter = 0;
+        while (this.hasJobsRemaining() && counter < maxCount) {
```

Comment:
log when max count was reached?

Reply:
This is relict of the original code. I kept it, but I have no idea what it was trying to achieve. In my opinion one shall either:
- _abort the serialization_ - in such case there is `executor.shutdownNow()` method
- _wait for all serializations to finish - in such case there is `executor.shutdown()`

One calls either first or other method. I have no idea why the _shutdown sequence_ is so complicated.